src/dosext/misc/emm.c

```diff
@@ -273,33 +273,44 @@ static void set_map_registers(const struct emm_reg *buf, int pages);
 #define Kdebug1(args)		E_Stub args
 #define Kdebug2(args)		E_Stub args
 
+static int ems_init_done = 0;
+
 void
 ems_helper(void) {
   switch (LWORD(ebx)) {
   case 0:
     E_printf("EMS Init called!\n");
     if (!config.ems_size) {
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_DISABLED_IN_CONFIG;
       return;
     }
+    if (ems_init_done) {
+      LWORD(eax) = 1;	/* report error */
+      LWORD(ebx) = EMS_ERROR_ALREADY_INITIALISED;
+      return;
+    }
     if (HI(ax) < DOSEMU_EMS_DRIVER_MIN_VERSION) {
       error("EMS driver version mismatch: got %i, expected %i, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("\nEMS driver version mismatch, disabling.\n"
             "Please update your ems.sys from the latest dosemu package.\n"
             "\nPress any key!\n");
+      LWORD(eax) = 1;	/* report error */
       LWORD(ebx) = EMS_ERROR_VERSION_MISMATCH;
       set_IF();
       com_biosgetch();
       clear_IF();
       return;
     }
+
     if (HI(ax) < DOSEMU_EMS_DRIVER_VERSION) {
       warn("EMS driver too old, consider updating %i->%i\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("EMS driver too old, consider updating.\n");
     }
+    ems_init_done = 1;	/* make sure the driver isn't loaded twice */
```

Comment:
Mmm, I'm not sure about that any more. With 
~~~
$_ems = (off)
~~~
I see message in DOS about EMS not being available, a message about dosemu XMS being installed and yet the driver and handler is not installed as the 'andb NoEMS, %al' is true and the jump to Error is done. It's as if the 'call HookHimem' is returning AL == 1, which I think means XMS not installed. I don't know enough about EMS/XMS DOS memory to say what the correct behaviour should be when we turn off EMS in dosemu.conf. Can you check the HookHimem call does the right thing in that case please?